hail/src/main/c/Makefile

```diff
@@ -39,14 +42,33 @@ OBJECTS := \
 
 BUILD_OBJECTS := $(OBJECTS:%=$(BUILD)/%.o)
 
-TEST_OBJECTS := $(foreach obj, $(OBJECTS), $(foreach file, $(wildcard $(obj)_test.cpp), $(BUILD)/$(basename $(file)).o))
+# before libsimdpp and catch.hpp are downloaded, clang -MG -MM will generate
+# unresolved dependencies
+.PHONY: simdpp/simd.h catch.hpp
+simdpp/simd.h: $(LIBSIMDPP)
+catch.hpp: $(CATCH_HEADER_LOCATION)
+
+TEST_CPP := $(foreach obj,$(OBJECTS),$(wildcard $(obj)_test.cpp))
+TEST_OBJECTS := $(foreach file,$(TEST_CPP),$(BUILD)/$(basename $(file)).o)
 TEST_OBJECTS += $(BUILD)/unit-tests.o
 
+HEADER_DEPENDENCIES := $(BUILD_OBJECTS:%.o=%.d) $(TEST_OBJECTS:%.o=%.d)
+-include $(HEADER_DEPENDENCIES)
+
 $(BUILD)/unit-tests.o: testutils/unit-tests.cpp
+	@mkdir -p $(@D)
 	$(CXX) -o $@ $(CXXFLAGS) -c $<
 
+$(BUILD)/%.d: %.cpp
+	@mkdir -p $(@D)
+	$(CXX) $(CXXFLAGS) $< -MG -M -MF $@ -MT $(@:%.d=%.o)
+
 $(BUILD)/%.o: %.cpp
-	$(CXX) -o $@ $(CXXFLAGS) -c $<
+	@mkdir -p $(@D)
+	$(CXX) -o $@ $(CXXFLAGS) -MD -MF $(@:%.o=%.d) -MT $@ -c $<
+
+$(BUILD)/ibs.d: CXXFLAGS += -DNUMBER_OF_GENOTYPES_PER_ROW=256
```

Comment:
More GCC problems! It appears that in the preprocessor for gcc < 8, it treats `(X % 0) != 0` as true‽ And still respects `#error` even in the presence of -MG.

```
$ CXX=g++-7 make build/ibs.d
g++-7 -march=sandybridge -O3 -std=c++14 -Ilibsimdpp-2.1 -Wall -Werror -fPIC -ggdb -fno-strict-aliasing -I../resources/include -I/usr/lib/jvm/java-8-openjdk/include -I/usr/lib/jvm/java-8-openjdk/include/linux -DNUMBER_OF_GENOTYPES_PER_ROW=256 ibs.cpp -MG -M -MF build/ibs.d -MT build/ibs.o
In file included from ibs.cpp:5:0:
ibs.h:36:46: error: division by zero in #if
 #if (NUMBER_OF_UINT64_GENOTYPE_PACKS_PER_ROW % UINT64_VECTOR_SIZE) != 0
                                              ^
ibs.h:37:2: error: #error "genotype packs per row, NUMBER_OF_UINT64_GENOTYPE_PACKS_PER_ROW, must be multuple of vector width, UINT64_VECTOR_SIZE."
 #error "genotype packs per row, NUMBER_OF_UINT64_GENOTYPE_PACKS_PER_ROW, must be multuple of vector width, UINT64_VECTOR_SIZE."
  ^~~~~
```
Potential workaround, add `-DHAIL_OVERRIDE_WIDTH=4` to just this rule

Reply:
oy vey

Reply:
I'm sorry I wrote `ibs.cpp` this is all my fault.